scripts/usb_scan.py

```diff
@@ -0,0 +1,753 @@
+            'bInterfaceClass': 'ff', 'bInterfaceProtocol': '00',
+            'bInterfaceSubClass': 'ff'})
+            |__ UsbInterface: (u'4-1.2:1.1', {'bInterfaceNumber': '01',
+            'bInterfaceClass': 'ff', 'bInterfaceProtocol': '01',
+            'bInterfaceSubClass': '42'})
+            |__ Allow: False
+            |__ Storages: []
+
+        :param device:(UsbDevice) the device to check
+        :return:(bool) if allow pass through
+        """
+        log_wrapper("policy check: " + device.get_node())
+        for i in device.get_all_interfaces():
+            allow_interface = False
+            for r in self.rule_list:
+                if self.match_device_interface(r, device, i):
+                    if r[self._ALLOW]:
+                        log_wrapper("allow " + i.get_node())
+                        allow_interface = True
+                        break
+                    else:
+                        log_wrapper("deny " + i.get_node())
+                        return False
+            if not allow_interface:
+                log_wrapper("deny " + i.get_node() + ", no matching rule")
+                return False
+
+        log_wrapper("allow " + device.get_node())
+        return True
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="scanner to get USB devices info")
+    parser.add_argument("-d", "--diagnostic", dest="diagnostic",
+                        action="store_true",
+                        help="enable diagnostic mode")
+    return parser.parse_args()
+
+
+def to_pusb(device):
+    """ convert UsbDevice to pusb dict
+
+    Example pusb dict:
+    [{"product-desc": "", "product-id": "383c", "description":
+    "Lenovo_H8RU2TGE", "vendor-desc": "Lenovo", "version": "2.00", "vendor-id":
+    "17ef", "path": "1-2", "serial": "H8RU2TGE"}, {"product-desc":
+    "TransMemory-Mini / Kingston DataTraveler 2.0 Stick (2GB)",
+    "product-id": "6544", "description":"Toshiba Corp._TransMemory-Mini /
+    Kingston DataTraveler 2.0 Stick (2GB)_9875B1B3F9F4CD301917964B",
+      "vendor-desc": "Toshiba Corp.", "version": "2.00", "vendor-id": "0930",
+      "path": "4-1.2", "serial": "9875B1B3F9F4CD301917964B"}]
+
+    :param device:(UsbDevice) the device to convert
+    :return:(dict) the key value pairs for pusb
+    """
+    pusb = {}
+
+    pusb["path"] = device.get_node()
+    # strip the leading space of "version"
+    pusb["version"] = device[UsbDevice._VERSION].strip()
+    pusb["vendor-id"] = device[UsbDevice._ID_VENDOR]
+    pusb["product-id"] = device[UsbDevice._ID_PRODUCT]
+    pusb["vendor-desc"] = device[UsbDevice._DESC_VENDOR]
+    pusb["product-desc"] = device[UsbDevice._DESC_PRODUCT]
+    pusb["serial"] = device[UsbDevice._SERIAL]
+
+    pusb["description"] = pusb["vendor-desc"]
+    if pusb["product-desc"]:
+        pusb["description"] += "_" + pusb["product-desc"]
+    if pusb["serial"]:
+        pusb["description"] += "_" + pusb["serial"]
+
```

Comment:
Hi @TalonsLee , I think we need to add strip to all the fields of PUSB.

Reply:
Why you want to strip it?

Reply:
done. comments showing the CA ticket added.